src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlConnection.cs

```diff
@@ -271,8 +288,49 @@ internal bool IsColumnEncryptionSettingEnabled
         /// <include file='../../../../../../../doc/snippets/Microsoft.Data.SqlClient/SqlConnection.xml' path='docs/members[@name="SqlConnection"]/RegisterColumnEncryptionKeyStoreProviders/*' />
         public static void RegisterColumnEncryptionKeyStoreProviders(IDictionary<string, SqlColumnEncryptionKeyStoreProvider> customProviders)
         {
+            ValidateCustomProviders(customProviders);
 
-            // Return when the provided dictionary is null.
+            lock (_GlobalCustomColumnEncryptionKeyProvidersLock)
+            {
+                // Provider list can only be set once
+                if (_GlobalCustomColumnEncryptionKeyStoreProviders != null)
+                {
+                    throw SQL.CanOnlyCallOnce();
+                }
+
+                // Create a temporary dictionary and then add items from the provided dictionary.
+                // Dictionary constructor does shallow copying by simply copying the provider name and provider reference pairs
+                // in the provided customerProviders dictionary.
+                Dictionary<string, SqlColumnEncryptionKeyStoreProvider> customColumnEncryptionKeyStoreProviders =
+                    new Dictionary<string, SqlColumnEncryptionKeyStoreProvider>(customProviders, StringComparer.OrdinalIgnoreCase);
+
+                // Set the dictionary to the ReadOnly dictionary.
+                _GlobalCustomColumnEncryptionKeyStoreProviders = 
+                    new ReadOnlyDictionary<string, SqlColumnEncryptionKeyStoreProvider>(customColumnEncryptionKeyStoreProviders);
+            }
+        }
+
+        /// <include file='../../../../../../../doc/snippets/Microsoft.Data.SqlClient/SqlConnection.xml' path='docs/members[@name="SqlConnection"]/RegisterColumnEncryptionKeyStoreProvidersOnConnection/*' />
+        public void RegisterColumnEncryptionKeyStoreProvidersOnConnection(IDictionary<string, SqlColumnEncryptionKeyStoreProvider> customProviders)
+        {
+            ValidateCustomProviders(customProviders);
+
+            // Create a temporary dictionary and then add items from the provided dictionary.
+            // Dictionary constructor does shallow copying by simply copying the provider name and provider reference pairs
+            // in the provided customerProviders dictionary.
+            Dictionary<string, SqlColumnEncryptionKeyStoreProvider> customColumnEncryptionKeyStoreProviders =
+                new Dictionary<string, SqlColumnEncryptionKeyStoreProvider>(customProviders, StringComparer.OrdinalIgnoreCase);
+
+            // Set the dictionary to the ReadOnly dictionary.
+            // This method can be called more than once. Re-registering a new collection will replace the 
+            // old collection of providers.
+            _CustomColumnEncryptionKeyStoreProviders =
+                new ReadOnlyDictionary<string, SqlColumnEncryptionKeyStoreProvider>(customColumnEncryptionKeyStoreProviders);
```

Comment:
If you use an `IReadOnlyDictionary` interface to define the `_CustomColumnEncryptionKeyStoreProviders` then you can directly assign the `customColumnEncryptionKeyStoreProviders` to it without creating a new object.

- same note for Netfx.